modules/python/python-source.c

```diff
@@ -0,0 +1,465 @@
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "python-source.h"
+#include "python-logmsg.h"
+#include "python-helpers.h"
+#include "logthrsource/logthrsourcedrv.h"
+#include "str-utils.h"
+#include "string-list.h"
+
+typedef struct _PythonSourceDriver PythonSourceDriver;
+
+struct _PythonSourceDriver
+{
+  LogThreadedSourceDriver super;
+
+  gchar *class;
+  GList *loaders;
+  GHashTable *options;
+
+  void (*post_message)(PythonSourceDriver *self, LogMessage *msg);
+
+  struct
+  {
+    PyObject *class;
+    PyObject *instance;
+    PyObject *run_method;
+    PyObject *request_exit_method;
+    PyObject *suspend_method;
+    PyObject *wakeup_method;
+  } py;
+};
+
+typedef struct _PyLogSource
+{
+  PyObject_HEAD
+  PythonSourceDriver *driver;
+} PyLogSource;
+
+static PyTypeObject py_log_source_type;
+
+
+void
+python_sd_set_class(LogDriver *s, gchar *filename)
+{
+  PythonSourceDriver *self = (PythonSourceDriver *) s;
+
+  g_free(self->class);
+  self->class = g_strdup(filename);
+}
+
+void
+python_sd_set_option(LogDriver *s, gchar *key, gchar *value)
+{
+  PythonSourceDriver *self = (PythonSourceDriver *) s;
+  gchar *normalized_key = __normalize_key(key);
+  g_hash_table_insert(self->options, normalized_key, g_strdup(value));
+}
+
+void
+python_sd_set_loaders(LogDriver *s, GList *loaders)
+{
+  PythonSourceDriver *self = (PythonSourceDriver *) s;
+
+  string_list_free(self->loaders);
+  self->loaders = loaders;
+}
+
+static const gchar *
+python_sd_format_stats_instance(LogThreadedSourceDriver *s)
+{
+  PythonSourceDriver *self = (PythonSourceDriver *) s;
+  static gchar persist_name[1024];
+
+  if (s->super.super.super.persist_name)
+    g_snprintf(persist_name, sizeof(persist_name), "python,%s", s->super.super.super.persist_name);
+  else
+    g_snprintf(persist_name, sizeof(persist_name), "python,%s", self->class);
+
+  return persist_name;
+}
+
+static void
+_ps_py_invoke_void_method_by_name(PythonSourceDriver *self, const gchar *method_name)
+{
+  _py_invoke_void_method_by_name(self->py.instance, method_name, self->class, self->super.super.super.id);
+}
+
+static gboolean
+_ps_py_invoke_bool_method_by_name_with_args(PythonSourceDriver *self, const gchar *method_name)
+{
+  return _py_invoke_bool_method_by_name_with_args(self->py.instance, method_name, self->options, self->class,
+                                                  self->super.super.super.id);
+}
+
+static void
+_ps_py_invoke_void_function(PythonSourceDriver *self, PyObject *func, PyObject *arg)
+{
+  return _py_invoke_void_function(func, arg, self->class, self->super.super.super.id);
+}
+
+static gboolean
+_py_invoke_init(PythonSourceDriver *self)
+{
+  return _ps_py_invoke_bool_method_by_name_with_args(self, "init");
+}
+
+static void
+_py_invoke_deinit(PythonSourceDriver *self)
+{
+  _ps_py_invoke_void_method_by_name(self, "deinit");
```

Comment:
I think it won't hurt to store `init`, `deinit`, ... as the `run_method`, because now this is going to print a debug method each time init/deinit is called, and not in startup. I would like to get a message about missing (even optional) methods when I start syslog-ng not when they happens to be called. (typo in function name hard to catch otherwise).